docs/rfcs/033-timeline-archive.md

```diff
@@ -0,0 +1,291 @@
+- Archived branches will not use any new/additional compression or de-duplication beyond what
+  is already implemented for image layers (zstd per page).
+- The pageserver will not "auto start" archived branches in response to page_service API requests: they
+  are only activated explicitly via the HTTP API.
+- We will not implement a total offload of archived timelines from safekeepers: their control file (small) will
+  remain on local disk, although existing eviction mechanisms will remove any segments from local disk.
+- We will not expose any prometheus metrics for archived timelines, or make them visible in any
+  detailed HTTP APIs.
+- A parent branch may not be archived unless all its children are.
+
+## Impacted Components
+
+pageserver, storage controller
+
+## Terminology
+
+**Archived**: a branch is _archived_ when an HTTP API request to archive it has succeeded: the caller
+may assume that this branch is now very cheap to store, although this may not be physically so until the
+branch proceeds to the offloaded state.
+
+**Active** branches are branches which are available for use by page_service clients, and have a relatively
+high cost due to consuming local storage.
+
+**Offloaded** branches are a subset of _archived_ branches, which have had their local state removed such
+that they now consume minimal runtime resources and have a cost similar to the cost of object storage.
+
+**Activate** (verb): transition from Archived to Active
+
+**Archive** (verb): transition from Active to Archived
+
+**Offload** (verb): transition from Archived to Offloaded
+
+**Offload manifest**: an object stored in S3 that describes timelines which pageservers do not load.
+
+**Warm up** (verb): operation done on an active branch, by downloading its active layers.  Once a branch is
+warmed up, good performance will be available to page_service clients.
+
+## Implementation
+
+### High level flow
+
+We may think of a timeline which is archived and then activated as proceeding through a series of states:
+
+```mermaid
+stateDiagram
+  [*] --> Active(warm)
+  Active(warm) --> Archived
+  Archived --> Offloaded
+  Archived --> Active(warm)
+  Offloaded --> Active(cold)
+  Active(cold) --> Active(warm)
+```
+
+These states are _not_ all stored as a single physical state on the timeline, but rather represent the combination
+of:
+- the timeline's lifecycle state: active or archived, stored in the timeline's index
+- its offload state: whether pageserver has chosen to drop local storage of the timeline and write it into the
+  manifest of offloaded timelines.
+- cache state (whether it's warm or cold).
+
+### Storage format changes
+
+There are two storage format changes:
+1. `index_part.json` gets a new attribute `state` that describes whether the timeline is to
+   be considered active or archived.
+2. A new tenant-level _offload manifest_ object `offload-v1.json` describes which timelines a tenant does not need to load
+   at startup.
```

Comment:
Feels like the new tenant manifest requires some protection from split brain scenarios.

Scenario: We have two storage controller managed pageservers. PS1 is the primary for our tenant shard and PS2 is the secondary location. Network partition between storcon and PS1 happens (PS1 still running and can read/write from/to S3). Storage controller marks PS1 offline and reschedules the tenant shard to PS2. PS1 does not receive the location config request to detach due to network partition. Both pageservers can update the tenant shard manifest resulting in UB.

Reply:
yeah... this is what I'm mulling in the alternatives section where I talk about having a generation on the offload manifest's key.  That would definitely work and be correct, I just haven't fully convinced myself that it's definitely the right answer, vs some other scheme, so am hand-waving a bit still.

Because we already have generations as a tool for this, we can bolt this on quite late in the project if necessary (e.g. start out with an un-generation'd manifest and then implement the generation part at the end before we go live with it).

Reply:
> Because we already have generations as a tool for this, we can bolt this on quite late in the project if necessary (e.g. start out with an un-generation'd manifest and then implement the generation part at the end before we go live with it).

That's a good project management idea, but it would be good to decide on how this is going to be achieved at the RFC stage.

Reply:
most def - it's my gate for taking this out of draft

Reply:
For simplicity, I've just added this to the "Storage format changes" section:
```
For split-brain protection, the manifest object will be written with a generation suffix, in the same way as
index_part objects are (see [generation numbers RFC](025-generation-numbers.md)).  This will add some complexity, but
give us total safety against two pageservers with the same tenant attached fighting over the object.  Existing code
for finding the latest generation and for cleaning up old generations (in the scrubber) will be generalized to cover
the manifest file.
```

If we decide to do split brain protection some other way, then fine, but the generation number is the bulletproof existing mechanism.